console: add console.count() and console.clear()

jasnell opened the pull request:
Both are simple utility functions defined by the WHATWG [console spec](https://console.spec.whatwg.org/).

Ref: https://github.com/nodejs/node/issues/12675

/cc @addaleax 

##### Checklist
- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-guidelines)

##### Affected core subsystem(s)
console

jasnell commented:
planning to tackle an implementation of the `console.group*()` functions separately once I figure out a reasonable approach

addaleax commented on test/parallel/test-console-clear.js:
```diff
@@ -0,0 +1,28 @@
+const common = require('../common');
+const assert = require('assert');
+
+const stdout_write = global.process.stdout.write;
```
Fwiw I think all the `global.`s here are redundant

addaleax commented on test/parallel/test-console-clear.js:
```diff
@@ -0,0 +1,28 @@
+// The sequence for moving the cursor to 0,0 and clearing screen down
+const check =
+  Buffer.from([0x1b, 0x5b, 0x31, 0x3b, 0x31, 0x48, 0x1b, 0x5b, 0x30, 0x4a])
+        .toString();
```
Just `check = '\u001b[1;1H\u001b[0J'` should be fine :)

addaleax commented on test/parallel/test-console-clear.js:
```diff
@@ -0,0 +1,28 @@
+
+process.on('exit', () => {
+  assert.strictEqual(buf, check);
+});
```
I think this doesn’t need to be wrapped in `.on('exit')`

jasnell commented:
@addaleax ... updated to fix those nits :-)

jasnell commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/7694/

vsemozhetbyt commented on doc/api/console.md:
````diff
@@ -167,6 +167,41 @@ console.assert(false, 'this message will print, but no error thrown');
+Maintains an internal counter specific to `label` and outputs to `stdout` the
+number of times `console.count([label])` has been called.
+
+```
````
FWIW: in `repl.md` there were ` ```js` in such cases and I've tried to save this approach to preserve highlighting with something [like this](https://github.com/nodejs/node/blame/b4fea2a3d62da5e2e3f90d7f2109f02f927f7174/doc/api/repl.md#L79-L80).

jasnell replied:
Thanks! :-)

YurySolovyov commented on lib/console.js:
```diff
@@ -54,6 +56,8 @@ function Console(stdout, stderr, ignoreErrors = true) {
   prop.value = createWriteErrorHandler(stderr);
   Object.defineProperty(this, '_stderrErrorHandler', prop);
 
+  this[kCounts] = {};
```
Why not use `Map` here? Won't that remove the restriction about symbols?
Or at least `Object.create(null)` to go straight into dictionary mode.

jasnell replied:
I've updated to `Object.create(null)`... the symbol restriction is actually there to approximate the behavior implemented by browsers. We can easily support `Symbol` if we wanted to.

jasnell commented:
Failures in CI are unrelated.

jasnell commented on lib/console.js:
```diff
@@ -166,6 +170,36 @@ Console.prototype.assert = function assert(expression, ...args) {
+};
+
+// Defined by: https://console.spec.whatwg.org/#table
+Console.prototype.table = Console.prototype.log;
```
Oh, yeah, just forgot about it. I need to add that to the documentation :-)

jasnell commented on doc/api/console.md:
```diff
@@ -167,6 +167,43 @@ console.assert(false, 'this message will print, but no error thrown');
+When `stdout` is a TTY, calling `console.clear()` will attempt to clear the
+TTY. When `stdout` is not a TTY, this method does nothing.
+
+### console.count([label])
```
O.o ... sigh. lol

Fishrock123 commented:
We shouldn't implement `console.clear()`. Not only does not not actually clear (You can actually clear terminals with hotkeys, but I don't think from stdio) but we aren't actually required to implement it by the spec. It doesn't make much sense in this context, ergo make it a no-op.

vsemozhetbyt commented on doc/api/console.md:
```diff
@@ -250,6 +290,15 @@ If formatting elements (e.g. `%d`) are not found in the first string then
+* `data` {any}
+* `...args` {any}
+
+Alias for `console.log([data[, ...args]]).
```
Is this an intentional way to designate optional parameters?

jasnell replied:
Yes.

vsemozhetbyt replied:
This is just the only such case in the doc and it differs from the very `console.log([data][, ...args])` fragment, but if it is intentional, it is OK)

jasnell replied:
oh, I see what you're referring to. No, that was just a misplaced `]` on my part :-)

jasnell replied:
Fixed!

vsemozhetbyt replied:
But it still miss a backtick)

jasnell replied:
That's fixed too :-)

jasnell commented:
@Fishrock123 
> does not actually clear

Not sure what you mean here. It clears just fine in the terminal.

Fishrock123 commented:
Doesn't that just jump down to make it look like there is no output in the screen? I investigated implementing this a while ago (might have been almost a year ago) and it didn't look like there was a way to actually clear it so there was no current terminal history loaded (e.g. no scroll up).

jasnell commented:
Sure, it leaves the scrollback buffer in place but that seems like a reasonable compromise.

TimothyGu commented on lib/console.js:
```diff
@@ -166,6 +170,36 @@ Console.prototype.assert = function assert(expression, ...args) {
+};
+
+// Defined by: https://console.spec.whatwg.org/#count
+Console.prototype.count = function count(label = kDefaultCount) {
```
According to the [IDL](https://console.spec.whatwg.org/#console-namespace), `label` is a DOMString with default being `'default'`. This means that the implementation should be akin to:

```js
function count(label = 'default') {
  label = `${label}`;
  // ...
  this.log(`${label}: ${counts[label]}`);
}
```

This should also allow the use of `Map` for `this[kCounts]`.

jasnell replied:
Yeah, I saw that but looking at the browser implementations, neither Chrome or Firefox treat `console.count()` and `console.count('default')` as being equivalent. They implement separate counters.

TimothyGu replied:
Fair enough, but I would still be happier if a) this deviation is documented in a comment, and b) `this[kCounts]` is a Map.

domenic replied:
We've filed bugs and are fixing Chrome and Firefox; patches should be landing within the week.

jasnell replied:
Given that, I've no problem with using `default` as the label.

@TimothyGu ... I'm curious why you want `this[kCounts]` to be a map. It would not have an impact on the ability to use `Symbol` as a key given the `toString()` restriction that would still exist. Using the dictionary mode object works just as well. Can you elaborate?

TimothyGu commented on lib/console.js:
```diff
@@ -166,6 +170,36 @@ Console.prototype.assert = function assert(expression, ...args) {
+  // get reset. Within the browser, that likely is not so much of a
+  // problem but it could be a source of nasty memory leaks in a
+  // Node.js process. We could add an option for clearing the counters
+  // but it would be non-standard.
```
That is indeed a problem. Can we file an issue at WHATWG, or supply a `console.countReset(label)` as an extension?

jasnell replied:
Opening an issue in the WHATWG repo would be good. I'd be hesitant to add a non-standard method before it was added to the spec.

domenic replied:
https://github.com/whatwg/console/issues/89

TimothyGu commented on doc/api/console.md:
```diff
@@ -250,6 +290,15 @@ If formatting elements (e.g. `%d`) are not found in the first string then
 [`util.inspect()`][] is called on each argument and the resulting string
 values are concatenated. See [`util.format()`][] for more information.
 
+### console.table([data][, ...args])
```
I'd prefer `console.table(tabularData[, properties])` like the [IDL](https://console.spec.whatwg.org/#console-namespace) specifies.

TimothyGu commented on doc/api/console.md:
```diff
@@ -250,6 +290,15 @@ If formatting elements (e.g. `%d`) are not found in the first string then
+* `data` {any}
+* `...args` {any}
+
+Alias for `console.log([data][, ...args])`.
```
This might be better implemented as `console.log(tabularData)`, since the second `properties` argument is not necessarily directly printed.

jasnell replied:
I'm contemplating just dropping this in this PR, to be honest. Still stewing on it tho.

jasnell commented:
@TimothyGu ... I've pushed a new commit that removes the `console.table()` alias.

@Fishrock123 ... I've added a note clarifying that `console.clear()` implementation may vary across OS and terminal impls.

not-an-aardvark commented on doc/api/console.md:
```diff
@@ -167,6 +167,52 @@ console.assert(false, 'this message will print, but no error thrown');
+-->
+
+* `label` {any} Any value (other than `Symbol`) that can be coerced into a
+  string.
```
Would it be clearer to just say `label` should be a string? Sure, the function will *work* if it's passed an object, but the output `[object Object]: 1` is usually not what a user would want. Additionally it might be confusing that two distinct objects share a label (since they will both be coerced to `[object Object]`).

jasnell replied:
This is fine but to match the behavior in browsers coercion will still be happening and two separate objects would still end up sharing a label unless they take steps to avoid it (e.g. by using `Symbol.toPrimitive`, etc)

lpinca commented on test/parallel/test-console-clear.js:
```diff
@@ -0,0 +1,22 @@
+const common = require('../common');
+const assert = require('assert');
+
+const stdout_write = process.stdout.write;
```
Nit: how about using camel case for the identifier?

lpinca commented on test/parallel/test-console-clear.js:
```diff
@@ -0,0 +1,22 @@
+// Fake TTY
+process.stdout.isTTY = true;
+
+process.stdout.write = common.mustCall((string) => buf += string, 2);
```
Nit: it doesn't harm, but I find `common.mustCall()` redundant in this case. The assertion below is sufficient imo.

Fishrock123 commented:
@jasnell I would prefer to not implement it until we can get clarification on the spec end what "clear" means. https://console.spec.whatwg.org/#clear is, unfortunately, not very clear about that.

jasnell commented:
@domenic ... any opinions on @Fishrock123's concerns around `console.clear()`?

domenic commented:
I don't understand them. The spec is pretty clear that you can do whatever is appropriate for your environment, including nothing. In general if something is about UI, it's up to the implementation to do what it thinks is best for its users.

jasnell commented:
Ok. Then let me rephrase the question :-) ... the current implementation I have here simply writes the escape codes to position the cursor at `0, 0` and clearing the screen down. The effect is the same as calling `clear` on most Linux terminals (the scrollback buffer is preserved). On Windows the effect is slightly different in that it clears the viewport for the running Node.js instance only (not the entire terminal screen) and there is no scrollback buffer to preserve. Is this, in your opinion, an acceptable implementation per the spec?

jasnell commented:
@domenic ... one additional question: should it be possible to use a `Symbol` as a counter label? Right now, the various implementations throw a TypeError because of the `toString()` limitation with `Symbol`s

domenic commented:
> Is this, in your opinion, an acceptable implementation per the spec?

Yes, definitely.

> should it be possible to use a Symbol as a counter label?

No, only strings.

jasnell commented:
Updated to address most recent comments:

* added `console.countReset()`, 
* simplified tests as suggested
* use `'default'` as the default label

cjihrig commented on lib/console.js:
```diff
@@ -166,6 +169,37 @@ Console.prototype.assert = function assert(expression, ...args) {
+Console.prototype.count = function count(label = 'default') {
+  const prefix = label !== 'default' ? `${label}: ` : '';
+  const counts = this[kCounts];
+  if (!counts[label])
```
Could you check against `undefined` here.

jasnell commented:
New CI: https://ci.nodejs.org/job/node-test-pull-request/7706/

jasnell commented:
@Fishrock123 .. given the response from @domenic do you still object?

Fishrock123 commented:
I'm not quite sure. What does `console.clear()` do in a pipe? Or to a file? Does it clear both stderr and stdout to a terminal, or do terminals not care for that? Does it work properly on Windows e.g. `CMD.exe`?

I personally have always strongly disliked terminal `clear` commands and always thought they were a pretty lazy attempt to do what they pretend to do (and seem to say they do). (I use `CMD+K` on macOS instead. There's a linux variant too but I forget it off-hand. Not sure about Windows.)

Fishrock123 commented:
_Aside: these should probably have been in separate PRs. I don't have anything to note about `console.count()` other than it adds an API that can be easily mimicked/polyfilled/module'd.)_

jasnell commented:
@Fishrock123 ... `console.clear()` only works if `isTTY` is `true`, in which case, all it does is write the escape codes to the stream, which is a best case effort (which is consistent with similar mechanisms in other environments). The escape codes are written only to `stdout` currently. It's up to the terminal implementation to determine how it is going to handle those escape codes. And yes, this works correctly on Windows (with both `CMD.exe` and PowerShell).

Fishrock123 commented:
Can we detect if the terminal accepts those escape codes? (I'm guessing probably not.)

addaleax commented:
> Can we detect if the terminal accepts those escape codes? (I'm guessing probably not.)

No, I don’t think that’ll work. :/

> `console.clear()` would need a `pseudo-tty` test though.

`console` isn’t actually concerned with I/O – I would personally prefer it if the tests here didn’t meddle with stdio at all and just used new `Console` instances with custom streams, that may or may not have the `isTTY` flag set.

Fishrock123 commented:
... That's not a good way of testing reality though ...

addaleax commented:
@Fishrock123 If you are going for integration tests, then message tests would be better anyway. You don’t really get very close to “reality” by just overriding `process.stdout.write` … it’s a good thing console and I/O are orthogonal, and for unit tests, imho it’s fine to use that orthogonality.

TimothyGu commented on doc/api/console.md:
```diff
@@ -167,6 +167,78 @@ console.assert(false, 'this message will print, but no error thrown');
+will clear only the output in the current terminal viewport for the Node.js
+binary.
+
+### console.count([label = 'default'])
```
I don't think the default parameter construction is supported by the documentation toolchain, and in any case it's unusual. The "defaults to `'default'`" description below should suffice.

TimothyGu commented on doc/api/console.md:
```diff
@@ -167,6 +167,78 @@ console.assert(false, 'this message will print, but no error thrown');
+* `label` {string} The display label for the counter. Defaults to `'default'`.
+
+Maintains an internal counter specific to `label` and outputs to `stdout` the
+number of times `console.count([label])` has been called. When the label is
```
This to me might look a bit like `label` is part of an array literal... How about

> `console.count()` has been called with the given `label`?

TimothyGu commented on doc/api/console.md:
````diff
@@ -167,6 +167,78 @@ console.assert(false, 'this message will print, but no error thrown');
+<!-- eslint-disable -->
+```js
+> console.count()
+1
````
If I'm reading the spec right, shouldn't this be `default: 1`?

jasnell replied:
Technically yes, browser implementations handle this differently tho. Not sure yet which we should go with. I'd be ok with `default: 1` tho.

domenic replied:
Again, we're working actively to converge everyone :)

TimothyGu commented on lib/console.js:
```diff
@@ -54,6 +55,8 @@ function Console(stdout, stderr, ignoreErrors = true) {
   prop.value = createWriteErrorHandler(stderr);
   Object.defineProperty(this, '_stderrErrorHandler', prop);
 
+  this[kCounts] = Object.create(null);
```
The reason why I preferred a Map is, well, this object _is_ used as a hashmap. We should try to prevent depending on how V8 internally implements object lookup but try to make the code [explicit/declarative](http://benediktmeurer.de/2017/04/03/v8-behind-the-scenes-march-edition/) &ndash; especially when a Map provides [performance improvements](https://github.com/nodejs/node/pull/11930) over `Object.create(null)` too.

jasnell commented:
Updated! Commits squashed... PTAL

jasnell commented:
@Fishrock123 ... no, there is no way of detecting if any particular terminal accepts the escape codes. Also, given that we do not appear to have any special pseudo-tty tests for any other API that outputs escape codes, I'm not sure why we would need them here. The `console.clear()` implementation is not doing anything special -- in fact, it's simply reusing public APIs from `readline` that have been in use for quite some time already.

(now that I look at it, there aren't even any tests for the `readline` APIs being used. This PR introduces what appears to be the first test of `readline.clearScreenDown()` and `readline.cursorTo()`

TimothyGu commented on lib/console.js:
```diff
@@ -166,6 +169,42 @@ Console.prototype.assert = function assert(expression, ...args) {
+// the counter from being a memory leak.
+Console.prototype.countReset = function countReset(label = 'default') {
+  const counts = this[kCounts];
+  counts.delete(label);
```
Does this need the string coercion treatment too?

jasnell replied:
No, I don't think so. It's an idempotent operation that can simply ignore invalid input.

TimothyGu replied:
For now I'm personally okay with that, but if this method gets spec'd, WHATWG will probably use the DOMString type. By then it'd be too hard to change the behavior in Node.js.

domenic replied:
Yes, this should be consistent with the other methods...

TimothyGu replied:
This should still be fixed.

jasnell commented:
Ping @Fishrock123 ... I'd like to get this landed.

Fishrock123 commented:
¯\\\_(ツ)\_/¯

jasnell commented:
Not sure what that means. You are the only one with a "requested changes" review.

jasnell commented:
Rebased.

jasnell commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/9339/

TimothyGu commented on test/parallel/test-console-clear.js:
```diff
@@ -0,0 +1,22 @@
+let buf = '';
+
+// Fake TTY
+process.stdout.isTTY = true;
```
This should probably (artificially) test the case where `isTTY = false` as well.

jasnell commented:
@TimothyGu ... updated! :-)

jasnell commented:
Landed in cc43c8f ...

fyi @addaleax ... this should cherry-pick cleanly into v8.x-staging if you wanted to pull it in to 8.3.0 at all. :-)

gibfahn commented:
Release team were +1 on backporting to v6.x.
